pkg/events/dep.go

```diff
@@ -0,0 +1,78 @@
+package events
+
+import (
+	"sync"
+	"sync/atomic"
+
+	"github.com/aquasecurity/tracee/pkg/ebpf/probes"
+)
+
+//
+// Dependencies (events, ksyms, tailcalls & capabilities)
+//
+
+//
+// NOTE:
+//
+// Within the Dependencies type, there is a single field which thread-safety is guaranteed by the
+// field type itself, and not by Dependencies methods: Capabilities. This happens because there are
+// currently two options to chose when adding fields to Dependencies:
+//
+// 1. To have a complex type (array, map, slice, ...) as a Dependencies field and protect it with
+//    mutexes that are also internal to Dependencies.
+//
+// 2. To have a single atomic pointer dereference to a type instance pointer, and rely on its
+//    internal synchronization mechanisms (like the Dependencies type).
+//
+
+type Dependencies struct {
+	events        map[ID]struct{}                 // map[eventID]struct{}
+	probes        map[probes.Handle]*Probe        // map[handle]*ProbeDependency
+	kSymbols      map[string]*KSymbol             // map[symbolName]*KSymbolDependency
+	tailCalls     map[string]map[string]*TailCall // map[mapName]map[progName]*TailCall
+	capabilities  *atomic.Pointer[Capabilities]   // pointer to a Capabilities instance
+	eventsLock    *sync.RWMutex
+	probesLock    *sync.RWMutex
+	kSymbolsLock  *sync.RWMutex
+	tailCallsLock *sync.RWMutex
```

Comment:
I agree BUT I don't know what will be mutable or not (yet) until the API server is designed and the extensions feature is finished. My syn here is to allow setters/getters for most of the fields and I'm not against disallowing that after the API server (and other features) are done. I needed something to start with and chose to allow everything first and reduce as appropriate (instead of the opposite).

Reply:
About tailcalls being mutable or not: I can think of multiple situations where adding/removing indexes would be done during runtime (during a reconfiguration of policies and traced events). Like I said, until our reconfiguration capabilities is finished we don't know what should be instantiable and what shouldn't.

Reply:
There is also something else I just thought.. by having event instances we can now move things out of Tracee singleton, such as eventConfig. The "submit/emit" fields are global and should be atomic and concurrently updated if/when needed.

Your "hack" for the tailcalls was also because we didn't have the "dynamic nature" of a "event definition" well defined. In your case, we started having instances in the definition, in the eventConfig case we started keeping track of things externally to the interface.

If you note, we had a "mix" of mutable vs immutable fields because we thought the definition of event would be "always static". By making it not "static" we can encapsulate the interface, better control concurrency and have a single point of view for atomic settings.

The event definition would become (and is, already, partially is) "event definition and status". WDYT ?